beacon-chain/das/availability.go

```diff
@@ -0,0 +1,250 @@
+// - BlobSidecars waiting for verification in the cache will be persisted to the db after verification.
+// - When BlobSidecars are written to the db, their cache entries are cleared.
+// - BlobSidecar cache entries with commitments that do not match the block will be evicted.
+// - BlobSidecar cachee entries with commitments that fail proof verification will be evicted.
+func (s *CachingDBVerifiedStore) IsDataAvailable(ctx context.Context, current primitives.Slot, b blocks.ROBlock) error {
+	blockCommitments := commitmentsToCheck(b, current)
+	if len(blockCommitments) == 0 {
+		return nil
+	}
+
+	key := keyFromBlock(b)
+	entry := s.cache.ensure(key)
+	// holding the lock over the course of the DA check simplifies everything
+	entry.Lock()
+	defer entry.Unlock()
+	if err := s.daCheck(ctx, b.Root(), blockCommitments, entry); err != nil {
+		return err
+	}
+	// If there is no error, DA has been successful, so we can clean up the cache.
+	s.cache.delete(key)
+	return nil
+}
+
+func (s *CachingDBVerifiedStore) daCheck(ctx context.Context, root [32]byte, blockCommitments [][]byte, entry *cacheEntry) error {
+	// retrieve BlobSidecars from the cache that match the block commitments. The order of blockCommitments is significant -
+	// the array offset of each [48]byte is assumed to correspond to the .Index field of the related BlobSidecar.
+	m, cachedScs := entry.filter(blockCommitments)
+	// The first return value from filter() is the number of items in the cache that match their corresponding commitments.
+	// If all BlobSidecars for the block are in the cache, we can avoid touching the database as a special case.
+	// This is usually the case when the cache is used for initial-sync or backfill, where caches can be ephemeral (one per batch).
+	if m == len(blockCommitments) {
+		if err := kzg.IsDataAvailable(blockCommitments, cachedScs); err == nil {
+			// We have all the committed sidecars in cache, and they all have valid proofs.
+			// If flushing them to backing storage succeeds, then we can confirm DA.
+			return s.db.SaveBlobSidecar(ctx, cachedScs)
+		}
+	}
+
+	// Since we don't have all the sidecars we need in cache, we'll try the more complicated check:
+	// merge the blobs in cache with those observed in the db. Try to verify all commitments together, and failing that,
+	// verify them one-by-one to find the set that match, and write the proven committed BlobSidecars back to the db.
+	// Anything already written to the db can be assumed correct - previously a block with the given root was observed
+	// that committed to those blobs, and their commitments to the data were proven.
+	dbidx, err := s.persisted(ctx, root, entry)
+	if err != nil {
+		return err
+	}
+	missing, err := dbidx.missing(blockCommitments)
+	// This is a database integrity sanity check - it should never fail.
+	if err != nil {
+		return err
+	}
+	// All commitments were found in the db, due to a previous successful DA check.
+	if len(missing) == 0 {
+		return nil
+	}
```

Comment:
Side note, during initial syncing, if we get a batch where `len(commitments) != len(sidecars)`, this is where we'll end up. If we dont end up here then there's a bug